src/protocol/sort/compose.rs

```diff
@@ -0,0 +1,127 @@
+pub struct Compose<'a, F> {
+    sigma: &'a mut Vec<Replicated<F>>,
+    rho: &'a mut Vec<Replicated<F>>,
+}
+
+impl<'a, F: Field> Compose<'a, F> {
+    #[allow(dead_code)]
+    pub fn new(sigma: &'a mut Vec<Replicated<F>>, rho: &'a mut Vec<Replicated<F>>) -> Self {
+        Self { sigma, rho }
+    }
+    #[embed_doc_image("compose", "images/sort/compose.png")]
+    /// This algorithm composes two permutations (`rho` and `sigma`). Both permutations are secret-shared,
+    /// and none of the helpers should learn it through this protocol.
+    /// Steps
+    /// ![Compose steps][compose]
+    /// 1. Generate random permutations using prss
+    /// 2. First permutation (sigma) is shuffled with random permutations
+    /// 3. Reveal the permutation
+    /// 4. Revealed permutation is applied locally on another permutation shares (rho)
+    /// 5. Unshuffle the permutation with random permutations to negate the effect of shuffling
+    #[allow(dead_code)]
+    pub async fn execute(&mut self, ctx: ProtocolContext<'_, F>) -> Result<(), BoxError> {
+        let mut random_permutations =
+            get_two_of_three_random_permutations(self.rho.len(), &ctx.prss());
+        let mut random_permutations_copy = random_permutations.clone();
+
+        let mut shuffle_sigma = Shuffle::new(self.sigma);
+        shuffle_sigma
+            .execute(ctx.narrow(&ShuffleSigma), &mut random_permutations)
+            .await?;
+
+        let mut perms = reveal_a_permutation(ctx.narrow(&RevealPermutation), self.sigma).await?;
+
+        apply_inv(&mut perms, &mut self.rho);
+
+        let mut unshuffle_rho = Shuffle::new(self.rho);
+        unshuffle_rho
+            .execute_unshuffle(ctx.narrow(&UnshuffleRho), &mut random_permutations_copy)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use permutation::Permutation;
+    use rand::seq::SliceRandom;
+    use tokio::try_join;
+
+    use crate::{
+        protocol::{
+            sort::{apply::apply_inv, compose::Compose},
+            QueryId,
+        },
+        test_fixture::{
+            generate_shares, make_contexts, make_world, validate_result_from_shares, TestWorld,
+        },
+    };
+
+    #[tokio::test]
+    pub async fn compose() {
+        const BATCHSIZE: usize = 25;
+        for _ in 0..10 {
+            let mut rng_sigma = rand::thread_rng();
+            let mut rng_rho = rand::thread_rng();
+
+            let mut sigma: Vec<usize> = (0..BATCHSIZE).collect();
+            sigma.shuffle(&mut rng_sigma);
+
+            let sigma_u128: Vec<u128> = sigma.iter().map(|x| *x as u128).collect();
+
+            let mut rho: Vec<usize> = (0..BATCHSIZE).collect();
+            rho.shuffle(&mut rng_rho);
+            let rho_u128: Vec<u128> = rho.iter().map(|x| *x as u128).collect();
+
+            let mut rho_composed = rho_u128.clone();
+
+            let mut sigma_copy = Permutation::oneline(sigma.clone());
+            apply_inv(&mut sigma_copy, &mut rho_composed);
```

Comment:
I think the less locals the better, because all the names just get confusing.
If it'll let you do this, I'd rather something like this:
```
apply_inv(
    &mut Permutation::oneline(sigma.clone()), 
    rho_u128.clone(),
);
```

Reply:
I still need the rho_composed to validate the results. So keeping that